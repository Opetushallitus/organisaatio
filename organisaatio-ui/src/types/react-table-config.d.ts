import {
    UseColumnOrderInstanceProps,
    UseExpandedInstanceProps,
    UseExpandedOptions,
    UseFiltersInstanceProps,
    UseFiltersOptions,
    UseGlobalFiltersInstanceProps,
    UseGlobalFiltersOptions,
    UseGroupByInstanceProps,
    UseGroupByOptions,
    UsePaginationInstanceProps,
    UsePaginationOptions,
    UseResizeColumnsOptions,
    UseRowSelectInstanceProps,
    UseRowSelectOptions,
    UseRowStateInstanceProps,
    UseRowStateOptions,
    UseSortByInstanceProps,
    UseSortByOptions,
} from 'react-table';

declare module 'react-table' {
    // take this file as-is, or comment out the sections that don't apply to your plugin configuration

    export interface TableOptions<D extends Record<string, unknown>>
        extends UseExpandedOptions<D>,
            UseFiltersOptions<D>,
            UseGlobalFiltersOptions<D>,
            UseGroupByOptions<D>,
            UsePaginationOptions<D>,
            UseResizeColumnsOptions<D>,
            UseRowSelectOptions<D>,
            UseRowStateOptions<D>,
            UseSortByOptions<D>,
            // note that having Record here allows you to add anything to the options, this matches the spirit of the
            // underlying js library, but might be cleaner if it's replaced by a more specific type that matches your
            // feature set, this is a safe default.
            Record<string, any> {}

    export interface TableInstance<D extends Record<string, unknown> = Record<string, unknown>>
        extends UseColumnOrderInstanceProps<D>,
            UseExpandedInstanceProps<D>,
            UseFiltersInstanceProps<D>,
            UseGlobalFiltersInstanceProps<D>,
            UseGroupByInstanceProps<D>,
            UsePaginationInstanceProps<D>,
            UseRowSelectInstanceProps<D>,
            UseRowStateInstanceProps<D>,
            UseSortByInstanceProps<D> {}

    export interface TableState<D extends Record<string, unknown> = Record<string, unknown>>
        extends UseColumnOrderState<D>,
            UseExpandedState<D>,
            UseFiltersState<D>,
            UseGlobalFiltersState<D>,
            UseGroupByState<D>,
            UsePaginationState<D>,
            UseResizeColumnsState<D>,
            UseRowSelectState<D>,
            UseRowStateState<D>,
            UseSortByState<D> {}
}
