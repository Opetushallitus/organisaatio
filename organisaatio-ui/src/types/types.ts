export type Language = 'fi' | 'sv' | 'en';
export type LocalDate = string;
export type LocalizableText = Partial<Record<Language, string>>;

// koodisto
export type KoodiUri = string;
export type KoodiArvo = string;
export type Koodi = {
    uri: KoodiUri;
    arvo: KoodiArvo;
    nimi: LocalizableText;
    versio: number;
};

// lokalisointi
export type Lokalisointi = Record<Language, Record<string, string>>;

export type KielistettyNimi = {
    nimi: string;
    kieli: Language; // ytj-kieli
    alkuPvm: LocalDate | null;
};

export type Osoite = {
    osoite?: string;
    postinumeroUri?: KoodiUri;
    postitoimipaikka?: string;
    email?: string;
    www?: string;
    numero?: string;
};

export type ytjOsoite = {
    katu: string;
    postinumero: string;
    toimipaikka: string;
    maa: string;
    kieli: boolean;
};

export type ytjYtunnus = {
    status?: string;
    alkupvm: LocalDate;
    loppupvm: LocalDate;
    yritysLopetettu: boolean;
    ytunnus: string;
};

type YhteystiedotBase = {
    id?: string;
    yhteystietoOid?: string;
    kieli: string;
};

type YhteystiedotEmail = YhteystiedotBase & {
    email: string;
};

type YhteystiedotPhone = YhteystiedotBase & {
    tyyppi: 'puhelin';
    numero: string;
};

type YhteystiedotWww = YhteystiedotBase & {
    www: string;
};

export type OsoiteType = 'posti' | 'kaynti' | 'ulkomainern_posti' | 'ulkomainen_kaynti' | 'muu';

export type YhteystiedotOsoite = YhteystiedotBase & {
    osoiteTyyppi: OsoiteType;
    postinumeroUri: KoodiUri;
    postitoimipaikka: string;
    osoite: string;
};

export type Yhteystiedot = YhteystiedotEmail | YhteystiedotPhone | YhteystiedotWww | YhteystiedotOsoite;

export type Nimi = {
    fi?: string;
    sv?: string;
    en?: string;
};
export type OrganisaatioBase = {
    oid: string;
    nimi: Nimi;
    status: string;
};
export type Organisaatio = OrganisaatioBase & {
    parentOid: string;
    parentOidPath: string;
    ytunnus: string;
    nimet: Nimi[];
    alkuPvm: LocalDate | null;
    yritysmuoto?: string;
    tyypit?: KoodiUri[];
    kotipaikkaUri: KoodiUri;
    muutKotipaikatUris?: KoodiUri[];
    maaUri?: KoodiUri;
    kieletUris: KoodiUri[];
    yhteystiedot?: Yhteystiedot[];
};
export type NewOrganisaatio = Omit<Organisaatio, 'oid'>;

export type NewRyhma = Omit<Ryhma, 'oid'>;

export interface YtjOrganisaatio {
    ytunnus: string;
    nimi: string;
    aloitusPvm: LocalDate | null;
    yritysmuoto: string;
    yritysmuotoKoodi: string;
    yritystunnusHistoria: any; // TODO?
    kayntiOsoite: ytjOsoite;
    kotiPaikka: string;
    kotiPaikkaKoodi?: string;
    postiOsoite: ytjOsoite;
    puhelin: string;
    toimiala: string;
    toimialaKoodi?: string;
    versio?: boolean;
    yrityksenKieli?: string;
    yritysTunnus: ytjYtunnus;
}

export type Ryhma = Omit<OrganisaatioBase, 'oid'> & {
    oid?: string;
    yritysmuoto?: string; // TODO Tuleeko n채m채???
    kuvaus?: any; // TODO Tuleeko n채m채???
    kayntiosoite?: any;
    kayttoryhmat: string[];
    kieletUris?: any[];
    kuvaus2: any;
    lisatiedot?: string[];
    lakkautusPvm?: string;
    muutKotipaikatUris?: string[];
    muutOppilaitosTyyppiUris?: string[];
    nimet?: any[];
    parentOid?: string;
    parentOidPath?: string;
    piilotettu?: boolean;
    postiosoite?: any;
    ryhmatyypit: string[];

    toimipistekoodi?: string;
    tyypit: string[];
    version?: number;
    vuosiluokat?: any[];
    yhteystiedot?: Yhteystiedot[];
    yhteystietoArvos?: any[];
};
export type OrganisaatioSuhde = {
    alkuPvm: string;
    loppuPvm?: string;
    child: OrganisaatioBase;
    parent: OrganisaatioBase;
};

export type YhdistaOrganisaatioon = {
    newParent?: string;
    date: Date;
    merge: boolean;
};

export type OrganisaatioHistoria = {
    childSuhteet: OrganisaatioSuhde[];
    parentSuhteet: OrganisaatioSuhde[];
    liitokset: OrganisaatioSuhde[];
    liittymiset: OrganisaatioSuhde[];
};
export interface YhteystietoTyyppi {
    allLisatietokenttas: any;
    oid?: string;
    nimi: any;
    sovellettavatOppilaitostyyppis: string[];
    sovellettavatOrganisaatios: string[];
    version: number;
}

export interface OrganisaatioNimiJaOid {
    oid: string;
    nimi: any;
}

export type SelectOptionType = {
    value: string;
    label: string;
};

export type TranslatedInputBind = {
    localizationKey: string;
    name: string;
    value: string;
    onChange: (e: React.FormEvent<HTMLInputElement>) => void;
    disabled?: boolean;
};

export type FrontProperties = {
    urlVirkailija: string;
};

export interface iOption {
    label: string;
    value: string;
}
