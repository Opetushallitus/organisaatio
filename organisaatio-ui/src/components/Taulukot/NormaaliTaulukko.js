import * as React from 'react';
import styles from './NormaaliTaulukko.module.css';
import {
  useExpanded, usePagination,
  useTable,
} from 'react-table';
import Button from "@opetushallitus/virkailija-ui-components/Button";
import {Icon} from "@iconify/react";
import chevronLeft from "@iconify/icons-fa-solid/chevron-left";
import chevronRight from "@iconify/icons-fa-solid/chevron-right";
import {useContext} from "react";
import {LanguageContext} from "../../contexts/contexts";



const mapPaginationSelectors = (index) => {
  if (index < 3) return [0, 5];
  return [index-2, index+3];
}

export default function NormaaliTaulukko({ data: inputData = [], tableColumns = [] }) {
    const { i18n, language } = useContext(LanguageContext);
    const columns = React.useMemo(
        () => tableColumns,
        []
    );
    const data = React.useMemo(
      () => inputData,
      []
    );
    console.log('pöö', inputData, tableColumns);
    const {
      getTableProps,
      getTableBodyProps,
      headerGroups,
      prepareRow,
      page, // Instead of using 'rows', we'll use page,
      // which has only the rows for the active page

      // The rest of these things are super handy, too ;)
      canPreviousPage,
      canNextPage,
      pageOptions,
      pageCount,
      gotoPage,
      nextPage,
      previousPage,
      setPageSize,
      state: { pageIndex, pageSize },
    } = useTable({ columns, data, initialState: { pageIndex: 0 } },
      useExpanded, usePagination);
      return(
        <div>
          <table {...getTableProps()} style={{ width: '100%', borderSpacing: 0 }}>
            <thead>
            {headerGroups.map(headerGroup => (
              <tr {...headerGroup.getHeaderGroupProps()}>
                {headerGroup.headers.map(column => (
                  <th
                    {...column.getHeaderProps({
                      className: column.collapse ? styles.collapse : '',
                    })}
                    style={{ textAlign: 'left', borderBottom: '1px solid rgba(151,151,151,0.5)'}}
                  >
                    {column.render('Header')}
                  </th>
                ))}
              </tr>
            ))}
            </thead>
            <tbody {...getTableBodyProps()}>
            {page.map((row, index) => {
              prepareRow(row);
              return (
                <tr {...row.getRowProps()}>
                  {row.cells.map(cell => {
                    return (
                      <td
                        {...cell.getCellProps({
                          className: cell.column.collapse ? styles.collapse : '',
                        })}
                        style={{
                          background: index % 2 === 0 ? '#F5F5F5': '#FFFFFF',
                        }}
                      >
                        {cell.render('Cell')}
                      </td>
                    )
                  })}
                </tr>
              )
            })}
            </tbody>
          </table>
          <div className={styles.PaginationContainer}>
            <div className={styles.PaginationSivunvaihto}>
              <Button variant='text' color="secondary" onClick={() => previousPage()} disabled={!canPreviousPage}>
                <Icon icon={chevronLeft} />
              </Button>
              {pageOptions.slice(...mapPaginationSelectors(pageIndex)).map(option => {
                if (option === pageIndex) return (<Button onClick={() => gotoPage(option)}>
                  {option+1}
                </Button>);
                return (<Button variant='text' color="secondary" onClick={() => gotoPage(option)}>
                  {option+1}
                </Button>);
              })}
              <Button variant="text" color="secondary" onClick={() => nextPage()} disabled={!canNextPage}>
                <Icon icon={chevronRight} />
              </Button>

            </div>
            <div className={styles.PaginationYhteensa}>
              <span>{i18n.translate('TAULUKKO_NAYTA_SIVULLA')}:</span>
              <select
                className={styles.NaytaSivullaSelect}
                value={pageSize}
                onChange={e => {
                  setPageSize(Number(e.target.value))
                }}
              >
                {[10, 20, 30, 40, 50].map(pageSize => (
                  <option key={pageSize} value={pageSize}>
                    {pageSize}
                  </option>
                ))}
              </select>
            </div>
          </div>
        </div>
      );
}