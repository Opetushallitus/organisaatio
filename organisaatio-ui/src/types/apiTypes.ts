import { KoodiUri, Language, LocalDate, Nimi, OrganisaationNimetNimi, OrganisaatioSuhde } from './types';

export type CASMe = {
    uid: string;
    oid: string;
    firstName: string;
    lastName: string;
    groups: string[];
    roles: string;
    lang: Language;
};
export type OrganisaatioLiitos = {
    alkuPvm: string;
    loppuPvm?: string;
    kohde: OrganisaatioBase;
    organisaatio: OrganisaatioBase;
};

export type APIOrganisaatioHistoria = {
    childSuhteet: OrganisaatioSuhde[];
    parentSuhteet: OrganisaatioSuhde[];
    liitokset: OrganisaatioLiitos[];
    liittymiset: OrganisaatioLiitos[];
};

type YhteystiedotBase = {
    id?: string;
    yhteystietoOid?: string;
    kieli: string;
    isNew?: boolean;
};

export type YhteystiedotEmail = YhteystiedotBase & {
    email: string;
};

export type YhteystiedotPhone = YhteystiedotBase & {
    tyyppi: 'puhelin';
    numero: string;
};

export type YhteystiedotWww = YhteystiedotBase & {
    www: string;
};

type OsoiteType = 'posti' | 'kaynti' | 'ulkomainen_posti' | 'ulkomainen_kaynti' | 'muu';

export type YhteystiedotOsoite = YhteystiedotBase & {
    osoiteTyyppi: OsoiteType;
    postinumeroUri: KoodiUri;
    postitoimipaikka: string;
    osoite: string;
};

export type ApiYhteystiedot = YhteystiedotEmail | YhteystiedotPhone | YhteystiedotWww | YhteystiedotOsoite;

export type OrganisaatioBase = {
    oid: string;
    status: string;
    nimi: Nimi;
    parentOid: string;
    parentOidPath: string;
};
export type ApiOrganisaatio = OrganisaatioBase & {
    alkuPvm: LocalDate;
    parentOid: string;
    parentOidPath: string;
    yritysmuoto?: string;
    tyypit: KoodiUri[];
    status: string;
    nimet: OrganisaationNimetNimi[];
    kotipaikkaUri: KoodiUri;
    muutKotipaikatUris?: KoodiUri[];
    maaUri: KoodiUri;
    kieletUris: KoodiUri[];
    yhteystiedot: ApiYhteystiedot[];
};

export type NewApiOrganisaatio = Omit<ApiOrganisaatio, 'oid' | 'status' | 'parentOidPath'>;
