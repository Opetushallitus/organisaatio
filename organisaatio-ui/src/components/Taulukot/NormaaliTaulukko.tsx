import * as React from 'react';
import { useState } from 'react';
import styles from './NormaaliTaulukko.module.css';
import {
    Column,
    FilterValue,
    HeaderGroup,
    Row,
    TableInstance,
    useExpanded,
    useFilters,
    useGlobalFilter,
    usePagination,
    useSortBy,
    useTable,
} from 'react-table';
import Button from '@opetushallitus/virkailija-ui-components/Button';
import Input from '@opetushallitus/virkailija-ui-components/Input';
import searchIcon from '@iconify/icons-fa-solid/search';
import Select from '@opetushallitus/virkailija-ui-components/Select';
import chevronLeft from '@iconify/icons-fa-solid/chevron-left';
import chevronRight from '@iconify/icons-fa-solid/chevron-right';
import { Ryhma, SelectOptionType } from '../../types/types';
import { ValueType } from 'react-select';
import IconWrapper from '../IconWapper/IconWrapper';
import { useAtom } from 'jotai';
import { languageAtom } from '../../api/lokalisaatio';
import { kayttoRyhmatKoodistoAtom, ryhmanTilaKoodistoAtom, ryhmaTyypitKoodistoAtom } from '../../api/koodisto';

const mapPaginationSelectors = (index: number) => {
    if (index < 3) return [0, 5];
    return [index - 2, index + 3];
};

export type NormaaliTaulukkoProps = {
    ryhmatData?: Ryhma[];
    ryhmatColumns?: Column<Ryhma>[];
    useHakuFiltteri?: boolean;
};

export type FiltteritProps = {
    setFilter: (columnId: string, updater: string | undefined | (string | null | undefined)[]) => void;
    setGlobalFilter: (filterValue: FilterValue) => void;
    globalFilter:
        | string
        | ((rows: Row<Ryhma>[], columnIds: string[], filterValue: unknown) => Row<Ryhma>[])
        | undefined;
};

export const chooseTaulukkoData = (ryhmatData: Ryhma[], ryhmatColumns: Column<Ryhma>[]) => {
    if (ryhmatData && ryhmatData.length > 0) {
        return { data: ryhmatData, columns: ryhmatColumns };
    }
    return { data: [], columns: [] };
};

export const Hakufiltterit = ({ setFilter, globalFilter, setGlobalFilter }: FiltteritProps) => {
    const [i18n] = useAtom(languageAtom);
    const [ryhmaTyypitKoodisto] = useAtom(ryhmaTyypitKoodistoAtom);
    const [kayttoRyhmatKoodisto] = useAtom(kayttoRyhmatKoodistoAtom);
    const [ryhmanTilaKoodisto] = useAtom(ryhmanTilaKoodistoAtom);
    const [tyyppiFiltteri, setTyyppiFiltteri] = useState<
        ValueType<SelectOptionType>[] | ValueType<SelectOptionType> | undefined
    >();
    const [kayttoRyhmatFiltteri, setKayttoRyhmatFiltteri] = useState<ValueType<SelectOptionType>[] | undefined>();
    const [tilaFiltteri, setTilaFiltteri] = useState<ValueType<SelectOptionType>[] | undefined>();

    const ryhmatyypitOptions = ryhmaTyypitKoodisto.selectOptions();
    const kayttoRyhmatOptions = kayttoRyhmatKoodisto.selectOptions();
    const ryhmanTilaOptions = ryhmanTilaKoodisto.selectOptions();

    return (
        <div>
            <div className={styles.FiltteriContainer}>
                <div className={styles.FiltteriRivi}>
                    <div className={styles.FiltteriInputOsa}>
                        <Input
                            placeholder={i18n.translate('RYHMAT_HAKU_PLACEHOLDER')}
                            onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                                setGlobalFilter(e.target.value || undefined); // Set undefined to remove the filter entirely
                            }}
                            value={globalFilter}
                            suffix={<IconWrapper color={'#999999'} icon={searchIcon} />}
                        />
                    </div>
                </div>
                <div className={styles.FiltteriRivi}>
                    <div className={styles.Kentta}>
                        <label>{i18n.translate('RYHMAT_RYHMAN_TYYPPI')}</label>
                        <Select
                            id={'RYHMAN_TYYPPI_SELECT'}
                            onChange={(
                                values: ValueType<SelectOptionType>[] | ValueType<SelectOptionType> | undefined
                            ) => {
                                setTyyppiFiltteri(values as ValueType<SelectOptionType>[] | undefined);
                                let tyyppi: string | undefined = undefined;
                                if (values && (values as ValueType<SelectOptionType>[]).length > 0) {
                                    tyyppi = (values as ValueType<SelectOptionType>[])
                                        .map((val) => val && (val as SelectOptionType).label)
                                        .join(', ');
                                }
                                setFilter('Tyyppi', tyyppi || undefined);
                            }}
                            isMulti
                            value={tyyppiFiltteri as ValueType<SelectOptionType>}
                            options={ryhmatyypitOptions}
                        />
                    </div>
                    <div className={styles.Kentta}>
                        <label>{i18n.translate('RYHMAT_RYHMAN_KAYTTOTARKOITUS')}</label>
                        <Select
                            id={'RYHMAN_KAYTTOTARKOITUS_SELECT'}
                            onChange={(
                                values: ValueType<SelectOptionType>[] | ValueType<SelectOptionType> | undefined
                            ) => {
                                setKayttoRyhmatFiltteri(values as ValueType<SelectOptionType>[] | undefined);
                                let kayttoryhmat: string | undefined = undefined;
                                if (values && (values as ValueType<SelectOptionType>[]).length > 0) {
                                    kayttoryhmat = (values as ValueType<SelectOptionType>[])
                                        .map((val) => val && (val as SelectOptionType).label)
                                        .join(', ');
                                }
                                setFilter('Kayttotarkoitus', kayttoryhmat || undefined);
                            }}
                            isMulti
                            value={kayttoRyhmatFiltteri as ValueType<SelectOptionType>}
                            options={kayttoRyhmatOptions}
                        />
                    </div>
                    <div className={styles.Kentta}>
                        <label>{i18n.translate('RYHMAT_RYHMAT_TILA')}</label>
                        <Select
                            id={'RYHMAN_TILA_SELECT'}
                            onChange={(
                                values: ValueType<SelectOptionType>[] | ValueType<SelectOptionType> | undefined
                            ) => {
                                setTilaFiltteri(values as ValueType<SelectOptionType>[] | undefined);
                                setFilter(
                                    'status',
                                    (values &&
                                        (values as ValueType<SelectOptionType>[]).map(
                                            (val) => val && (val as SelectOptionType).label
                                        )) ||
                                        undefined
                                );
                            }}
                            isMulti
                            value={tilaFiltteri as ValueType<SelectOptionType>}
                            options={ryhmanTilaOptions}
                        />
                    </div>
                </div>
            </div>
        </div>
    );
};

const NormaaliTaulukko = ({ ryhmatData = [], ryhmatColumns = [], useHakuFiltteri = false }: NormaaliTaulukkoProps) => {
    const [i18n] = useAtom(languageAtom);

    const { data, columns } = chooseTaulukkoData(ryhmatData, ryhmatColumns);
    const {
        getTableProps,
        getTableBodyProps,
        headerGroups,
        prepareRow,
        page, // Instead of using 'rows', we'll use page,
        // which has only the rows for the active page

        globalFilter,
        setGlobalFilter,
        setFilter,
        canPreviousPage,
        canNextPage,
        pageOptions,
        gotoPage,
        nextPage,
        previousPage,
        setPageSize,
        state: { pageIndex, pageSize },
    } = useTable(
        {
            columns,
            data,
            initialState: {
                pageIndex: 0,
                sortBy: [
                    {
                        id: 'Nimi',
                        desc: false,
                    },
                ],
            },
        },
        useFilters,
        useGlobalFilter,
        useSortBy,
        useExpanded,
        usePagination
    ) as TableInstance<Ryhma>;
    return (
        <div>
            {useHakuFiltteri && (
                <div>
                    <Hakufiltterit
                        globalFilter={globalFilter}
                        setGlobalFilter={setGlobalFilter}
                        setFilter={setFilter}
                    />
                </div>
            )}
            <table {...getTableProps()} style={{ width: '100%', borderSpacing: 0 }}>
                <thead>
                    {headerGroups.map((headerGroup) => (
                        <tr {...headerGroup.getHeaderGroupProps()} key={headerGroup.getHeaderGroupProps().key}>
                            {headerGroup.headers.map((column) => (
                                <th
                                    {...column.getHeaderProps({
                                        className: (column as HeaderGroup<Ryhma> & { collapse: boolean }).collapse
                                            ? styles.collapse
                                            : '',
                                    })}
                                    key={column.getHeaderProps().key}
                                    style={{ textAlign: 'left', borderBottom: '1px solid rgba(151,151,151,0.5)' }}
                                >
                                    {column.render('Header')}
                                </th>
                            ))}
                        </tr>
                    ))}
                </thead>
                <tbody {...getTableBodyProps()}>
                    {page.map((row, index) => {
                        prepareRow(row);
                        return (
                            <tr {...row.getRowProps()} key={row.getRowProps().key}>
                                {row.cells.map((cell) => {
                                    return (
                                        <td
                                            {...cell.getCellProps({
                                                className: (cell.column as HeaderGroup<Ryhma> & { collapse: boolean })
                                                    .collapse
                                                    ? styles.collapse
                                                    : '',
                                            })}
                                            key={cell.getCellProps().key}
                                            style={{
                                                background: index % 2 === 0 ? '#F5F5F5' : '#FFFFFF',
                                            }}
                                        >
                                            {cell.render('Cell')}
                                        </td>
                                    );
                                })}
                            </tr>
                        );
                    })}
                </tbody>
            </table>
            <div className={styles.PaginationContainer}>
                <div className={styles.PaginationSivunvaihto}>
                    <Button variant={'text'} color={'secondary'} onClick={previousPage} disabled={!canPreviousPage}>
                        <IconWrapper icon={chevronLeft} />
                    </Button>
                    {pageOptions.slice(...mapPaginationSelectors(pageIndex)).map((option) => {
                        if (option === pageIndex)
                            return (
                                <Button key={option} onClick={() => gotoPage(option)}>
                                    {option + 1}
                                </Button>
                            );
                        return (
                            <Button key={option} variant={'text'} color={'secondary'} onClick={() => gotoPage(option)}>
                                {option + 1}
                            </Button>
                        );
                    })}
                    <Button variant={'text'} color={'secondary'} onClick={nextPage} disabled={!canNextPage}>
                        <IconWrapper icon={chevronRight} />
                    </Button>
                </div>
                <div className={styles.PaginationYhteensa}>
                    <span>{i18n.translate('TAULUKKO_NAYTA_SIVULLA')}:</span>
                    <select
                        className={styles.NaytaSivullaSelect}
                        value={pageSize}
                        onChange={(e) => {
                            setPageSize(Number(e.target.value));
                        }}
                    >
                        {[10, 20, 30, 40, 50].map((pageSize) => (
                            <option key={pageSize} value={pageSize}>
                                {pageSize}
                            </option>
                        ))}
                    </select>
                </div>
            </div>
        </div>
    );
};

export default NormaaliTaulukko;
