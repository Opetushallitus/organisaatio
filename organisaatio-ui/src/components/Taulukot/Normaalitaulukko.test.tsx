import React from 'react';
import { shallow } from 'enzyme';
import '@testing-library/jest-dom/extend-expect';
import NormaaliTaulukko, {
    chooseTaulukkoData,
    FiltteritProps,
    Hakufiltterit,
    NormaaliTaulukkoProps,
} from './NormaaliTaulukko';
import { Ryhma } from '../../types/types';
import { Column } from 'react-table';
import { useAtom } from 'jotai';

const MINIMAL_PROPS: NormaaliTaulukkoProps = {
    ryhmatData: [],
    ryhmatColumns: [],
    useHakuFiltteri: false,
};

const MINIMAL_FILTTERIT_PROPS: FiltteritProps = {
    setFilter: jest.fn(),
    setGlobalFilter: jest.fn(),
    globalFilter: '',
};

jest.mock('jotai');
beforeEach(() => {
    jest.resetAllMocks();
    jest.spyOn(console, 'error').mockImplementation(() => {});
    (useAtom as jest.Mock).mockReturnValue([{ translate: (a) => a, translateNimi: (a) => a, koodit: () => [] }]);
});

afterAll(() => {
    jest.clearAllMocks();
});

describe('Normaalitaulukko', () => {
    describe('Rendering', () => {
        it('Renders without crashing', () => {
            const element = shallow(<NormaaliTaulukko {...MINIMAL_PROPS} />);
            expect(element).toMatchSnapshot();
        });
    });

    describe('Hakufiltterit', () => {
        it('Renders without crashing', () => {
            const element = shallow(<Hakufiltterit {...MINIMAL_FILTTERIT_PROPS} />);
            expect(element).toMatchSnapshot();
        });
        it('Is not rendered when useHakuFiltteri is set false', () => {
            const element = shallow(<NormaaliTaulukko {...{ ...MINIMAL_PROPS }} />);
            expect(element.find(Hakufiltterit)).toHaveLength(0);
        });
        it('Is rendered when useHakuFiltteri is set true', () => {
            const element = shallow(<NormaaliTaulukko {...{ ...MINIMAL_PROPS, useHakuFiltteri: true }} />);
            expect(element.find(Hakufiltterit)).toHaveLength(1);
        });
    });
    describe('chooseTaulukkoData', () => {
        const ryhma = {
            kayntiosoite: undefined,
            kayttoryhmat: [],
            kieletUris: [],
            kuvaus: '',
            kuvaus2: undefined,
            lakkautusPvm: '',
            lisatiedot: [],
            muutKotipaikatUris: [],
            muutOppilaitosTyyppiUris: [],
            nimet: [],
            nimi: undefined,
            oid: '',
            parentOid: '',
            parentOidPath: '',
            piilotettu: false,
            postiosoite: undefined,
            ryhmatyypit: [],
            status: '',
            toimipistekoodi: '',
            tyypit: [],
            version: 0,
            vuosiluokat: [],
            yhteystiedot: [],
            yhteystietoArvos: [],
            yritysmuoto: '',
        };

        const ryhmatColumn = {
            Header: 'nimi',
        } as Column<Ryhma>;

        test.each([
            [[ryhma], [ryhmatColumn], [], [], { data: [ryhma], columns: [ryhmatColumn] }],

            [[], [], [], [], { data: [], columns: [] }],
        ])(
            'Returns ryhmatData and ryhmatColumns or yhteystietotyypitData and yhteystietotyypitColumns based on which one has data',
            (ryhmatData, ryhmatColumns, yhteystietotyypitData, yhteystietotyypitColumns, expectedResult) => {
                const result = chooseTaulukkoData(ryhmatData, ryhmatColumns);
                expect(result).toEqual(expectedResult);
            }
        );
    });
});
