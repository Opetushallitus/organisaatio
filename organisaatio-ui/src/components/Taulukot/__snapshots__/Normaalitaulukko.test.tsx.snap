// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Normaalitaulukko Hakufiltterit Renders without crashing 1`] = `
<div>
  <div
    className="FiltteriContainer"
  >
    <div
      className="FiltteriRivi"
    >
      <div
        className="FiltteriInputOsa"
      >
        <ForwardRef
          onChange={[Function]}
          placeholder="RYHMAT_HAKU_PLACEHOLDER"
          suffix={
            <ForwardRef(Icon)
              color="#999999"
              icon={
                Object {
                  "body": "<path d=\\"M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128c0-70.7 57.2-128 128-128c70.7 0 128 57.2 128 128c0 70.7-57.2 128-128 128z\\" fill=\\"currentColor\\"/>",
                  "height": 512,
                  "width": 512,
                }
              }
            />
          }
          value=""
        />
      </div>
    </div>
    <div
      className="FiltteriRivi"
    >
      <div
        className="Kentta"
      >
        <label>
          RYHMAT_RYHMAN_TYYPPI
        </label>
        <Select
          id="RYHMAN_TYYPPI_SELECT"
          isMulti={true}
          onChange={[Function]}
          options={Array []}
        />
      </div>
      <div
        className="Kentta"
      >
        <label>
          RYHMAT_RYHMAN_KAYTTOTARKOITUS
        </label>
        <Select
          id="RYHMAN_KAYTTOTARKOITUS_SELECT"
          isMulti={true}
          onChange={[Function]}
          options={Array []}
        />
      </div>
      <div
        className="Kentta"
      >
        <label>
          RYHMAT_RYHMAT_TILA
        </label>
        <Select
          id="RYHMAN_TILA_SELECT"
          isMulti={true}
          onChange={[Function]}
          options={Array []}
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Normaalitaulukko Rendering Renders without crashing 1`] = `
<div>
  <table
    role="table"
    style={
      Object {
        "borderSpacing": 0,
        "width": "100%",
      }
    }
  >
    <thead />
    <tbody
      role="rowgroup"
    />
  </table>
  <div
    className="PaginationContainer"
  >
    <div
      className="PaginationSivunvaihto"
    >
      <ForwardRef
        color="secondary"
        disabled={true}
        onClick={[Function]}
        variant="text"
      >
        <ForwardRef(Icon)
          icon={
            Object {
              "body": "<path d=\\"M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z\\" fill=\\"currentColor\\"/>",
              "height": 512,
              "width": 320,
            }
          }
        />
      </ForwardRef>
      <ForwardRef
        color="secondary"
        disabled={true}
        onClick={[Function]}
        variant="text"
      >
        <ForwardRef(Icon)
          icon={
            Object {
              "body": "<path d=\\"M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256L34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z\\" fill=\\"currentColor\\"/>",
              "height": 512,
              "width": 320,
            }
          }
        />
      </ForwardRef>
    </div>
    <div
      className="PaginationYhteensa"
    >
      <span>
        TAULUKKO_NAYTA_SIVULLA
        :
      </span>
      <select
        className="NaytaSivullaSelect"
        onChange={[Function]}
        value={10}
      >
        <option
          key="10"
          value={10}
        >
          10
        </option>
        <option
          key="20"
          value={20}
        >
          20
        </option>
        <option
          key="30"
          value={30}
        >
          30
        </option>
        <option
          key="40"
          value={40}
        >
          40
        </option>
        <option
          key="50"
          value={50}
        >
          50
        </option>
      </select>
    </div>
  </div>
</div>
`;
